nupic/research/frameworks/dynamic_sparse/gmp_lr_scheduler.py

```diff
@@ -0,0 +1,119 @@
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+# See the GNU Affero Public License for more details.
+#
+# You should have received a copy of the GNU Affero Public License
+# along with this program.  If not, see http://www.gnu.org/licenses.
+#
+# http://numenta.org/licenses/
+# ----------------------------------------------------------------------
+
+from torch.optim.lr_scheduler import LambdaLR
+
+__all__ = [
+    "ThreeStageGMPLR",
+]
+
+
+class ThreeStageGMPLR(LambdaLR):
+    """
+    This is an lr schedule for Gradual Magnitude Pruning (GMP) consisting of three
+    stages:
+        1) warmup - linearly ramp up from 0 to max_lr; this differs from the paper
+                    which uses a constant lr
+        2) pruning - maintain a constant max_lr
+        3) cooldown - decay the learning rate twice (like a StepLR)
+
+    :param max_lr: this is the maximum lr reached
+    :param warmup_steps: number of steps for the warmup phase
+    :param cooldown_steps: number of steps for the cooldown phase
+    :param total_steps: total number of steps; used to derive pruning steps
+    :param cooldown_gamma: how much to decay the lr during cooldown
+                           e.g. lr <- lr * cooldown_gamma; used to decay the lr twice
+    """
+
+    def __init__(
+        self,
+        optimizer,
+        max_lr,
+        warmup_steps,
+        cooldown_steps,
+        total_steps,
+        cooldown_gamma
+    ):
+        self.max_lr = max_lr
+        self.warmup_steps = warmup_steps
+        self.cooldown_steps = cooldown_steps
+        self.total_steps = total_steps
+        self.pruning_steps = total_steps - warmup_steps - cooldown_steps
+        self.cooldown_gamma = 0.1
+        assert int(cooldown_steps / 3) >= 2, (
+            "Too few `cooldown_steps`. The lr will be decayed twice throughout the "
+            "cooldown period."
+        )
+
+        super().__init__(optimizer, self.composed_lr)
+
+    def composed_lr(self, step: int):
+        """
+        LR schedule for GMP Pruning consisting of three phases. Here `step` starts at 0.
+        """
+
+        # Step starts at 0, but will make it start at 1.
+        train_step = step + 1
+
+        # Phase 1: Warm-up phase - linear warmup.
+        if train_step < self.warmup_steps:
+            return self.warmup_lr(step)
+
+        # Phase 3: Cool-down phase - StepLR like decay
+        elif train_step > self.total_steps - self.cooldown_steps:
+            return self.cooldown_lr(step)
+
+        # Phase 2: Pruning phase - constant lr.
+        else:
+            return self.pruning_lr(step)
+
+    # Warm-up phase.
+    def warmup_lr(self, step: int):
+        """Linearly ramp up the lr from 0 to max_lr in `warmup_steps`"""
+        min_lr = 0
+        return (self.max_lr - min_lr) / (self.warmup_steps - 1) * step + min_lr
```

Comment:
Why set min_lr to zero every time? If min_lr is 0, why have + or - min_lr on line 92? If there is ever a case where min_lr is not 0, maybe set that as an attribute? Also, not crucial, but it looks to me like the first part of that expression `(self.max_lr - min_lr) / (self.warmup_steps - 1)` is just an increment rate, which can be computed once in advance. Then you could do something like 
`self.warmup_rate = (self.max_lr - self.min_lr) / (self.warmup_steps - 1)`
in the init, followed by
`return self.min_lr + (self.warmup_rate * step)`
in the warmup_lr function.

Reply:
Similar comment, maybe min_lr should be a class attribute that can be defined by the user.